File: app/src/main/java/com/apps/darkstorm/swrpg/assistant/swElements/Weapon.kt
package com.apps.darkstorm.swrpg.assistant.swElements

import android.util.JsonReader
import android.util.JsonToken
import android.util.JsonWriter
import com.apps.darkstorm.swrpg.assistant.saveLoad.JsonSavable

data class Weapon(var name: String = "",var damage: Int = 0,var critical: Int = 0,var hp: Int = 0,var range: Int = 0,
                  var skill: Int = 0, var skillBase: Int = 0,var characteristics: MutableList<WeaponCharacteristic> = mutableListOf(),
                  var addBrawn: Boolean = false, var loaded: Boolean = true,var limitedAmmo: Boolean = false,
                  var itemState: Int = 0, var ammo: Int = 0, var firingArc: String = "", var encumbrance: Int = 0):JsonSavable(){
    override fun save(jw: JsonWriter) {
        jw.beginObject()
        jw.name("name").value(name)
        jw.name("damage").value(damage)
        jw.name("critical rating").value(critical)
        jw.name("hard points").value(hp)
        jw.name("range").value(range)
        jw.name("skill").value(skill)
        jw.name("base").value(skillBase)
        jw.name("Weapon Characteristics").beginArray()
        for(wc in characteristics)
            wc.save(jw)
        jw.endArray()
        jw.name("add brawn").value(addBrawn)
        jw.name("loaded").value(loaded)
        jw.name("limited ammo").value(limitedAmmo)
        jw.name("item state").value(itemState)
        jw.name("ammo").value(ammo)
        jw.name("firing arc").value(firingArc)
        jw.name("encumbrance").value(encumbrance)
        jw.endObject()
    }
    override fun load(jr: JsonReader) {
        jr.beginObject()
        while(jr.hasNext() && jr.peek() != JsonToken.END_OBJECT){
            if(jr.peek()!= JsonToken.NAME){
                jr.skipValue()
                continue
            }
            val jName = jr.nextName()
            when(jName){
                "name"-> name = jr.nextString()
                "damage"-> damage = jr.nextInt()
                "critical rating"-> critical = jr.nextInt()
                "hard points"-> hp = jr.nextInt()
                "range"-> range = jr.nextInt()
                "skill"-> skill = jr.nextInt()
                "base"-> skillBase = jr.nextInt()
                "Weapon Characteristics"->{
                    characteristics = mutableListOf()
                    jr.beginArray()
                    while(jr.hasNext()){
                        val chr = WeaponCharacteristic()
                        chr.load(jr)
                        characteristics.add(chr)
                    }
                    jr.endArray()
                }
                "add brawn"-> addBrawn = jr.nextBoolean()
                "loaded"-> loaded = jr.nextBoolean()
                "limited ammo"-> limitedAmmo = jr.nextBoolean()
                "item state"-> itemState = jr.nextInt()
                "ammo"-> ammo = jr.nextInt()
                "firing arc"-> firingArc = jr.nextString()
                "encumbrance"-> encumbrance = jr.nextInt()
            }
        }
        jr.endObject()
    }
    override fun clone() = copy()
}

File: lib/items/weapon.dart

import 'package:flutter/material.dart';
import 'package:swassistant/sw.dart';
import 'package:swassistant/utils/json_savable.dart';

import 'weapon_characteristic.dart';

class Weapon implements JsonSavable{

  String name = "";
  int? damage;
  int? critical;
  int hp = 0;
  //0-Engaged, 1-Short, 2-Medium, 3-Long, 4-Extreme
  int range = 0;
  //0-Brawl, 1-Gunner, 2-Lightsaber, 3-Melee, 4-Ranged(Heavy), 5-Ranged(Light)
  int? skill;
  int? skillBase;
  List<WeaponCharacteristic> characteristics = [];
  bool addBrawn = false;
  bool loaded = true;
  bool limitedAmmo = false;
  //0-None, 1-Minor, 2-Major, 3-Maor
  int itemState = 0;
  int ammo = 0;
  String firingArc = "";
  int encumbrance = 0;

  Weapon();

  Weapon.fromJson(Map<String,dynamic> json) :
    name = json["name"] ?? "",
    damage = json["damage"],
    critical = json["critical rating"],
    hp = json["hard points"] ?? 0,
    range = json["range"] ?? 0,
    skill = json["skill"],
    skillBase = json["base"],
    addBrawn = json["add brawn"] ?? false,
    loaded = json["loaded"] ?? true,
    limitedAmmo = json["limited ammo"] ?? false,
    itemState = json["item state"] ?? 0,
    ammo = json["ammo"] ?? 0,
    firingArc = json["firing arc"] ?? "",
    encumbrance = json["encumbrance"] ?? 0,
    characteristics = []{
      if(json["Weapon Characteristics"] != null){
        for(Map<String,dynamic> map in json["Weapon Characteristics"]){
          characteristics.add(WeaponCharacteristic.fromJson(map));
        }
      }
    }

  Weapon.from(Weapon from) : 
    name = from.name,
    damage = from.damage,
    critical = from.critical, 
    hp = from.hp,
    range = from.range,
    skill = from.skill,
    skillBase = from.skillBase,
    addBrawn = from.addBrawn,
    loaded = from.loaded,
    limitedAmmo = from.limitedAmmo,
    itemState = from.itemState,
    ammo = from.ammo,
    firingArc = from.firingArc,
    encumbrance = from.encumbrance,
    characteristics = List.from(from.characteristics);

  @override
  Map<String,dynamic> toJson() => {
    "name" : name,
    "damage" : damage,
    "critical rating" : critical,
    "hard points" : hp,
    "range" : range,
    "skill" : skill,
    "base" : skillBase,
    "Weapon Characteristics" : List.generate(characteristics.length, (index) => characteristics[index].toJson()),
    "add brawn" : addBrawn,
    "loaded" : loaded,
    "limited ammo" : limitedAmmo,
    "item state" : itemState,
    "ammo" : ammo,
    "firing arc" : firingArc,
    "encumbrance" : encumbrance
  }..removeWhere((key, value) {
    if (value is List && value.isEmpty) return true;
    return zeroValue[key] == value;
  });

  @override
  Map<String, dynamic> get zeroValue => {
    "name": "",
    "hard points": 0,
    "range": 0,
    "add brawn": false,
    "loaded": true,
    "limited ammo": false,
    "item state": 0,
    "ammo": 0,
    "firing arc": "",
    "encumbrance": 0,
  };

  @override
  operator ==(other) => other is Weapon && other.name == name && other.damage == damage && other.critical == critical &&
      other.hp == hp && other.range == range && other.skill == skill && other.skillBase == skillBase &&
      other.addBrawn == addBrawn && other.loaded == loaded && other.limitedAmmo == limitedAmmo &&other.itemState == itemState &&
      other.ammo == ammo && other.firingArc == firingArc && other.encumbrance == encumbrance && other.characteristics == characteristics;


  @override
  int get hashCode => Object.hash(name, damage, critical, hp, range, skill, skillBase, addBrawn, loaded, limitedAmmo, itemState, ammo, firingArc, encumbrance, Object.hashAll(characteristics));

  static List<String> weaponSkills(BuildContext context) => [
    SW.of(context).locale.skills3,
    SW.of(context).locale.skills11,
    SW.of(context).locale.skills13,
    SW.of(context).locale.skills16,
    SW.of(context).locale.skills21,
    SW.of(context).locale.skills22,
  ];
}